### PR TITLE
modify transaction process

### PR DESCRIPTION
**Pull request by mask-pp**

### CODE REVIEW

File: cluster/master/peer.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ package master
 import (
 	"errors"
 	"fmt"
+	"github.com/QuarkChain/goquarkchain/cluster/rpc"
```
Comment (pingke) [inline cluster/master/peer.go]: move below

File: cluster/master/peer.go [hunk 2]
```diff
@@ -197,11 +193,8 @@ func (p *Peer) PeerID() string {
 
 // SendTransactions sends transactions to the peer and includes the hashes
 // in its transaction hash set for future reference.
-func (p *Peer) SendTransactions(branch uint32, txs []*types.Transaction) error {
-	data := p2p.NewTransactionList{}
-	data.TransactionList = txs
-
-	msg, err := p2p.MakeMsg(p2p.NewTransactionListMsg, 0, p2p.Metadata{Branch: branch}, data)
+func (p *Peer) SendTransactions(p2pTxs *rpc.P2PRedirectRequest) error {
+	msg, err := p2p.MakeMsgBytes(p2p.NewTransactionListMsg, 0, p2p.Metadata{Branch: p2pTxs.Branch}, p2pTxs.Data)
```
Comment (pingke) [inline cluster/master/peer.go]: MakeMsgWithSerializedData

File: cluster/slave/slave_grpc.go [hunk 3]
```diff
@@ -4,13 +4,14 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/QuarkChain/goquarkchain/params"
```
Comment (pingke) [inline cluster/slave/slave_grpc.go]: move below

File: p2p/p2p_commands.go [hunk 4]
```diff
@@ -72,14 +72,27 @@ func (p P2PCommandOp) String() string {
 	return reflect.TypeOf(OPSerializerMap[p]).Name()
 }
 
-func MakeMsg(op P2PCommandOp, rpcID uint64, metadata Metadata, msg interface{}) (Msg, error) {
+func MakeMsgBytes(op P2PCommandOp, rpcID uint64, metadata Metadata, msg []byte) (Msg, error) {
```
Comment (pingke) [inline p2p/p2p_commands.go]: rename

File: p2p/p2p_commands.go [hunk 5]
```diff
@@ -72,14 +72,27 @@ func (p P2PCommandOp) String() string {
...
 	return Msg{Code: 0, Size: uint32(len(qkcBody)), Payload: bytes.NewReader(qkcBody)}, nil
 }
 
+func MakeMsg(op P2PCommandOp, rpcID uint64, metadata Metadata, msg interface{}) (Msg, error) {
+	cmdBytes, err := serialize.SerializeToBytes(msg)
+	if err != nil {
+		return Msg{}, err
+	}
+
+	qkcBody, err := Encrypt(metadata, op, rpcID, cmdBytes)
```
Comment (pingke) [inline p2p/p2p_commands.go]: use MakeMsgBytes

File: cluster/slave/slave_grpc.go [hunk 6]
```diff
@@ -604,41 +605,67 @@ func (s *SlaveServerSideOp) HandleNewTip(ctx context.Context, req *rpc.Request)
...
 
 func (s *SlaveServerSideOp) AddTransactions(ctx context.Context, req *rpc.Request) (*rpc.Response, error) {
-	ts := time.Now()
 	var (
-		gReq     rpc.NewTransactionList
-		gRes     rpc.HashList
-		response = &rpc.Response{RpcId: req.RpcId}
-		err      error
+		gReq   rpc.P2PRedirectRequest
+		txsMsg p2p.NewTransactionList
```
Comment (pingke) [inline cluster/slave/slave_grpc.go]: txs

File: cluster/slave/api_backend.go [hunk 7]
```diff
@@ -156,26 +156,28 @@ func (s *SlaveBackend) AddTx(tx *types.Transaction) (err error) {
...
-	fromShardSize, err := s.clstrCfg.Quarkchain.GetShardSizeByChainId(tx.EvmTx.FromChainID())
-	if err != nil {
-		return nil, err
+	shard, ok := s.shards[branch]
+	if !ok {
+		return fmt.Errorf("fullShardID:%v not found", branch)
 	}
-	if err := tx.EvmTx.SetFromShardSize(fromShardSize); err != nil {
-		return nil, err
+	errList := shard.MinorBlockChain.AddTxList(txs)
```
Comment (pingke) [inline cluster/slave/api_backend.go]: errList?

Comment (mask-pp) [reply cluster/slave/api_backend.go]: yes, the same as before.

File: cluster/slave/slave_grpc.go [hunk 8]
```diff
@@ -604,41 +605,67 @@ func (s *SlaveServerSideOp) HandleNewTip(ctx context.Context, req *rpc.Request)
...
+		fullShardIds = s.slave.fullShardList
+	)
+	for _, tx := range txs.TransactionList {
+		fId := tx.EvmTx.FromFullShardId()
+		for _, id := range fullShardIds {
+			if fId != id {
+				continue
+			}
+			if _, ok := txList[id]; !ok {
+				txList[id] = make([]*types.Transaction, 0, params.NEW_TRANSACTION_LIST_LIMIT)
```
Comment (pingke) [inline cluster/slave/slave_grpc.go]: make([]*types.Transaction, 0,len(txs))

File: cluster/slave/slave_grpc.go [hunk 9]
```diff
@@ -604,41 +605,67 @@ func (s *SlaveServerSideOp) HandleNewTip(ctx context.Context, req *rpc.Request)
...
 
-	if len(errList) != len(gReq.TransactionList) {
-		return nil, errors.New("errList != txList")
-	}
-	for index, tx := range gReq.TransactionList {
-		if errList[index] == nil {
-			gRes.Hashes = append(gRes.Hashes, tx.Hash())
+	addTxList := func(branch uint32, txs []*types.Transaction) error {
+		ts := time.Now()
+		if len(txs) > params.NEW_TRANSACTION_LIST_LIMIT {
```
Comment (pingke) [inline cluster/slave/slave_grpc.go]: this can move outside

**Pull request merged**